manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,244 @@
+	i := 2
+	for _, resourceDeallocator := range resourceDeallocators {
+		events[i] = resourceDeallocator.event
+		i++
+	}
+
+	eventsChan, eventsChanCancel, err := store.ViewAndWatch(deallocator.store,
+		func(readTx store.ReadTx) error {
+			// look for services that are marked for deletion
+			// there's no index on the `PendingDelete` field,
+			// so we just iterate over all of them and filter manually
+			// this is okay since we only do this at leadership change
+			allServices, err := store.FindServices(readTx, store.All)
+
+			if err != nil {
+				log.G(ctx).WithError(err).Error("failed to list services in deallocator init")
+				return err
+			}
+
+			// let's populate our internal taskCounts
+			for _, service := range allServices {
+				if service.PendingDelete {
+					if err = deallocator.processPendingDeleteService(ctx, readTx, service); err != nil {
+						return err
+					}
+				}
+			}
+
+			// now we also need to look at all existing service-level resources
+			// that are marked for deletion
+			for _, resourceDeallocator := range resourceDeallocators {
+				allResources, err := resourceDeallocator.enumerator(readTx)
+
+				if err != nil {
+					log.G(ctx).WithError(err).Errorf("failed to list %vs in deallocator init", resourceDeallocator.name)
+					return err
+				}
+
+				for _, resource := range allResources {
+					if resource.isPendingDelete() {
+						if err := deallocator.maybeDeallocateResource(ctx, nil, resourceDeallocator, resource.id(), nil); err != nil {
+							return err
+						}
+					}
+				}
+			}
+
+			return nil
+		},
+		events...)
+
+	if err != nil {
+		// if we have an error here, we can't proceed any further, let the world burn
+		log.G(ctx).WithError(err).Error("failed to initialize the deallocator")
+		return err
+	}
+
+	defer func() {
+		close(deallocator.doneChan)
+		eventsChanCancel()
+	}()
+
+	// now we just need to wait for events
+	for {
+		select {
+		case event := <-eventsChan:
+			if err = deallocator.processNewEvent(ctx, event); err != nil {
+				log.G(ctx).WithError(err).Errorf("error processing deallocator event %#v", event)
+			}
+		case <-deallocator.stopChan:
+			return nil
+		}
+	}
+}
+
+func (deallocator *Deallocator) processPendingDeleteService(ctx context.Context, readTx store.ReadTx, service *api.Service) error {
+	var (
+		tasks []*api.Task
+		err   error
+	)
+
+	viewFunc := func(r store.ReadTx) {
+		tasks, err = store.FindTasks(r, store.ByServiceID(service.ID))
+	}
+
+	if readTx == nil {
+		deallocator.store.View(viewFunc)
+	} else {
+		viewFunc(readTx)
+	}
+
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to retrieve the list of tasks for service %v", service.ID)
+		// if in doubt, let's proceed to clean up the service anyway
+		// better to clean up resources that shouldn't be cleaned up yet
+		// than ending up with a service and some resources lost in limbo forever
+		return deallocator.deallocateService(ctx, service)
```

Comment:
this may cause problems. Its probably better to be safe and maybe retry.

Reply:
I thought about it. Sure, I'm not opposed to re-trying, but that only delays the real problem: what happens after we're retried n times and keep failing? Do we never stop trying (in which case this could be dangerous at another level, namely resource consumption & performance)? Or do we still forget after a while?